File: tools/competitor_research.py
"""
competitor_research.py
Söker upp och analyserar konkurrenter via Perplexity API.
Output: .tmp/research_raw.json
"""

import os
import json
import sys
import requests
from datetime import datetime
from pathlib import Path

# ── Sökvägar ────────────────────────────────────────────────────────────────
ROOT = Path(__file__).parent.parent
BRAND_FILE = ROOT / "brand" / "brand.json"
OUTPUT_FILE = ROOT / ".tmp" / "research_raw.json"
ENV_FILE = ROOT / ".env"

# ── Ladda .env manuellt (undviker beroende av python-dotenv) ────────────────
def load_env():
    if ENV_FILE.exists():
        for line in ENV_FILE.read_text(encoding="utf-8").splitlines():
            line = line.strip()
            if line and not line.startswith("#") and "=" in line:
                key, _, val = line.partition("=")
                os.environ.setdefault(key.strip(), val.strip())

# ── Perplexity API-anrop ────────────────────────────────────────────────────
def ask_perplexity(query: str, api_key: str) -> str:
    """Skickar en fråga till Perplexity och returnerar svaret som text."""
    url = "https://api.perplexity.ai/chat/completions"
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": "sonar",
        "messages": [
            {
                "role": "system",
                "content": (
                    "Du är en affärsanalytiker som specialiserar sig på den svenska digitala byrå-marknaden. "
                    "Svara alltid på svenska. Var konkret, faktabaserad och strukturerad. "
                    "Fokusera på verifierbara fakta om svenska byråer."
                ),
            },
            {"role": "user", "content": query},
        ],
        "max_tokens": 2000,
        "temperature": 0.2,
        "search_recency_filter": "month",
        "return_citations": True,
    }
    response = requests.post(url, headers=headers, json=payload, timeout=60)
    response.raise_for_status()
    data = response.json()
    return data["choices"][0]["message"]["content"]


# ── Queries ─────────────────────────────────────────────────────────────────
def build_queries(brand: dict) -> list[dict]:
    services_str = ", ".join(brand.get("services", []))
    target = brand.get("target_market", "kliniker i Sverige")
    return [
        {
            "id": "landscape",
            "label": "Marknadsöversikt",
            "query": (
                f"Lista de 10 mest framstående digitala byråerna i Sverige som erbjuder {services_str} "
                f"till {target}. För varje byrå: namn, webbplats, ungefärliga priser, "
                f"specialitet och styrkor. Fokusera på byråer aktiva 2024-2025."
            ),
        },
        {
            "id": "pricing",
            "label": "Prisanalys",
            "query": (
                f"Vad kostar det att anlita en digital byrå i Sverige för {services_str}? "
                "Ge konkreta prisintervall (SEK) för: "
                "1) Hemsida/landningssida (engångskostnad), "
                "2) SEO per månad, "
                "3) Google Ads management per månad, "
                "4) Paketpris allt-i-ett. "
                "Basera på verkliga svenska byråer 2024-2025."
            ),
        },
        {
            "id": "positioning",
            "label": "Positionering & gaps",
            "query": (
                f"Analysera hur svenska digitala byråer positionerar sig på marknaden för {target}. "
                "Vilka positioneringsstrategier används? "
                "Vilka nischer är underservade? "
                "Vad saknar kunderna hos befintliga byråer? "
                "Vilka differentieringsfaktorer är viktigast?"
            ),
        },
        {
            "id": "clinic_niche",
            "label": "Klinik-nischen specifikt",
            "query": (
                "Vilka digitala byråer i Sverige specialiserar sig på kliniker, skönhetssalonger "
                "och estetiska kliniker? Finns det byråer som fokuserar specifikt på denna nisch? "
                "Vad är deras erbjudande och priser? "
                "Hur marknadsför sig estetiska kliniker online i Sverige 2024-2025?"
            ),
        },
        {
            "id": "opportunities",
            "label": "Möjligheter",
            "query": (
                f"Vad är de största möjligheterna för en ny digital byrå i Sverige som fokuserar på "
                f"{target} och erbjuder {services_str}? "
                "Vilka konkurrensfördelar är möjliga? "
                "Vad efterfrågar kliniker och skönhetskliniker online? "
                "Hur kan en byrå differentiera sig och ta marknadsandelar?"
            ),
        },
    ]


# ── Huvud ───────────────────────────────────────────────────────────────────
def main():
    load_env()

    api_key = os.environ.get("PERPLEXITY_API_KEY", "")
    if not api_key or api_key == "din_nyckel_här":
        print("❌ Ingen Perplexity API-nyckel hittad.")
        print("   Öppna .env och fyll i: PERPLEXITY_API_KEY=din_nyckel_här")
        print("   Skaffa nyckel på: https://www.perplexity.ai/settings/api")
        sys.exit(1)

    if not BRAND_FILE.exists():
        print(f"❌ Hittade inte brand.json på: {BRAND_FILE}")
        sys.exit(1)

    brand = json.loads(BRAND_FILE.read_text(encoding="utf-8"))
    print(f"✅ Laddat brand: {brand.get('company_name')}")

    queries = build_queries(brand)
    results = {
        "generated_at": datetime.now().isoformat(),
        "brand": brand,
        "research": {},
    }

    for i, q in enumerate(queries, 1):
        print(f"\n[{i}/{len(queries)}] Researchar: {q['label']}...")
        try:
            answer = ask_perplexity(q["query"], api_key)
            results["research"][q["id"]] = {
                "label": q["label"],
                "query": q["query"],
                "answer": answer,
            }
            print(f"   ✅ Klart ({len(answer)} tecken)")
        except requests.HTTPError as e:
            print(f"   ❌ HTTP-fel: {e}")
            results["research"][q["id"]] = {
                "label": q["label"],
                "query": q["query"],
                "answer": f"FEL: {e}",
            }
        except Exception as e:
            print(f"   ❌ Oväntat fel: {e}")
            results["research"][q["id"]] = {
                "label": q["label"],
                "query": q["query"],
                "answer": f"FEL: {e}",
            }

    OUTPUT_FILE.parent.mkdir(exist_ok=True)
    OUTPUT_FILE.write_text(json.dumps(results, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"\n✅ Research sparad till: {OUTPUT_FILE}")
    print("   Kör nu: python tools/generate_report.py")


if __name__ == "__main__":
    main()

File: tools/competitor_research.js
#!/usr/bin/env node
/**
 * competitor_research.js
 * Söker konkurrenter via Perplexity API + SerpAPI för social media data.
 * Output: .tmp/research_raw.json
 */

import { readFileSync, writeFileSync, mkdirSync, existsSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, "..");
const BRAND_FILE = join(ROOT, "brand", "brand.json");
const OUTPUT_FILE = join(ROOT, ".tmp", "research_raw.json");
const ENV_FILE = join(ROOT, ".env");

// ── .env ─────────────────────────────────────────────────────────────────────
function loadEnv() {
  if (!existsSync(ENV_FILE)) return;
  for (const line of readFileSync(ENV_FILE, "utf-8").split("\n")) {
    const t = line.trim();
    if (!t || t.startsWith("#") || !t.includes("=")) continue;
    const [k, ...v] = t.split("=");
    process.env[k.trim()] ??= v.join("=").trim();
  }
}

// ── Perplexity ────────────────────────────────────────────────────────────────
async function askPerplexity(query, apiKey) {
  const res = await fetch("https://api.perplexity.ai/chat/completions", {
    method: "POST",
    headers: { Authorization: `Bearer ${apiKey}`, "Content-Type": "application/json" },
    body: JSON.stringify({
      model: "sonar",
      messages: [
        {
          role: "system",
          content: "Du är en affärsanalytiker specialiserad på den svenska digitala byrå-marknaden. Svara på svenska. Var konkret och faktabaserad med specifika namn, siffror och exempel.",
        },
        { role: "user", content: query },
      ],
      max_tokens: 2500,
      temperature: 0.2,
      search_recency_filter: "month",
    }),
  });
  if (!res.ok) throw new Error(`Perplexity HTTP ${res.status}: ${await res.text()}`);
  return (await res.json()).choices[0].message.content;
}

// ── SerpAPI ───────────────────────────────────────────────────────────────────
async function serpSearch(query, apiKey, num = 10) {
  const params = new URLSearchParams({
    q: query,
    api_key: apiKey,
    hl: "sv",
    gl: "se",
    num: String(num),
  });
  const res = await fetch(`https://serpapi.com/search?${params}`);
  if (!res.ok) throw new Error(`SerpAPI HTTP ${res.status}`);
  const data = await res.json();
  return (data.organic_results || []).map(r => ({
    title: r.title,
    link: r.link,
    snippet: r.snippet,
  }));
}

// ── Hämta Instagram/social via SerpAPI ──────────────────────────────────────
async function getSocialResults(apiKey) {
  const queries = [
    "svenska webbyrå digital byrå instagram followers kliniker estetik",
    'site:instagram.com "webbyrå" OR "digital byrå" Sverige kliniker',
    "webbbyrå SEO reklam Sverige instagram linkedin konto",
  ];
  const allResults = [];
  for (const q of queries) {
    try {
      console.log(`   SerpAPI: "${q.slice(0, 50)}..."`);
      const results = await serpSearch(q, apiKey, 10);
      allResults.push(...results);
    } catch (e) {
      console.warn(`   SerpAPI varning: ${e.message}`);
    }
  }
  // Dedup by link
  const seen = new Set();
  return allResults.filter(r => {
    if (seen.has(r.link)) return false;
    seen.add(r.link); return true;
  });
}

// ── Competitor search via SerpAPI ─────────────────────────────────────────────
async function getCompetitorResults(apiKey) {
  const queries = [
    "digital byrå hemsidor SEO reklam kliniker Sverige",
    "webbyrå estetisk klinik skönhetsklinik hemsida Sverige",
    "SEO byrå Google Ads hälsa klinik Stockholm Göteborg",
  ];
  const allResults = [];
  for (const q of queries) {
    try {
      console.log(`   SerpAPI: "${q.slice(0, 50)}..."`);
      const results = await serpSearch(q, apiKey, 10);
      allResults.push(...results);
    } catch (e) {
      console.warn(`   SerpAPI varning: ${e.message}`);
    }
  }
  const seen = new Set();
  return allResults.filter(r => {
    if (seen.has(r.link)) return false;
    seen.add(r.link); return true;
  });
}

// ── Perplexity queries ────────────────────────────────────────────────────────
function buildQueries(brand) {
  const services = (brand.services || []).join(", ");
  const target = brand.target_market || "kliniker i Sverige";

  return [
    {
      id: "landscape",
      label: "Marknadsöversikt",
      query: `Lista de 10 mest framstående digitala byråerna i Sverige som erbjuder ${services} till ${target}. För varje byrå: namn, webbplats, specialitet, styrkor och ungefärliga priser. Fokus 2024-2025.`,
    },
    {
      id: "pricing",
      label: "Prisanalys",
      query: `Vad kostar det att anlita en digital byrå i Sverige för ${services}? Konkreta prisintervall (SEK) för: 1) Hemsida/landningssida, 2) SEO per månad, 3) Google Ads management, 4) Paketpris. Nämn byrånamn. 2024-2025.`,
    },
    {
      id: "clinic_niche",
      label: "Klinik-nischen",
      query: `Vilka digitala byråer i Sverige specialiserar sig på kliniker, skönhetssalonger och estetiska kliniker? Nämn specifika byråer, deras erbjudande och priser. Hur marknadsför estetiska kliniker sig online i Sverige 2025?`,
    },
    {
      id: "positioning",
      label: "Positionering & gaps",
      query: `Analysera positioneringen hos svenska digitala byråer för ${target}. Vilka nischer är underservade? Vad saknar kunderna? Vilka differentieringsfaktorer är viktigast för en ny byrå 2025?`,
    },
    {
      id: "social_media",
      label: "Social Media & Konkurrenter",
      query: `Lista svenska digitala byråer och webbbyråer med aktiva Instagram- eller LinkedIn-konton med minst 1000 följare. För varje: kontonamn, plattform, ungefärligt följarantal, typ av innehåll de postar och vilken strategi de verkar använda. Fokus på byråer som riktar sig mot hälsa, kliniker eller B2B 2024-2025.`,
    },
    {
      id: "competitor_offers",
      label: "Konkurrenternas erbjudanden",
      query: `Vad säljer svenska webbbyråer och digitala byråer konkret till kliniker och hälsoföretag? Lista specifika paket, prismodeller (engång vs månadsavgift), vad som ingår och vilka resultat de lovar. Nämn byrånamn. Sverige 2024-2025.`,
    },
    {
      id: "opportunities",
      label: "Möjligheter",
      query: `Vilka är de största möjligheterna för en ny digital byrå i Sverige fokuserad på ${target} med tjänsterna ${services}? Vilka konkurrensfördelar är realistiska? Hur kan man ta marknadsandelar 2025?`,
    },
  ];
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  loadEnv();

  const perplexityKey = process.env.PERPLEXITY_API_KEY || "";
  const serpKey = process.env.SERPAPI_KEY || "";

  if (!perplexityKey || perplexityKey === "din_nyckel_här") {
    console.error("❌ Ingen Perplexity API-nyckel i .env");
    process.exit(1);
  }

  const brand = JSON.parse(readFileSync(BRAND_FILE, "utf-8"));
  console.log(`✅ Brand: ${brand.company_name}`);
  console.log(`✅ SerpAPI: ${serpKey ? "aktiv" : "saknas (hoppar över Google-sökningar)"}`);

  const results = {
    generated_at: new Date().toISOString(),
    brand,
    research: {},
    serp_results: {},
  };

  // SerpAPI: social media + competitor searches
  if (serpKey) {
    console.log("\n📡 Söker via SerpAPI...");
    try {
      console.log(" [1/2] Social media-konton...");
      results.serp_results.social = await getSocialResults(serpKey);
      console.log(`   ✅ ${results.serp_results.social.length} resultat`);
    } catch (e) {
      console.error(`   ❌ ${e.message}`);
      results.serp_results.social = [];
    }
    try {
      console.log(" [2/2] Konkurrenter...");
      results.serp_results.competitors = await getCompetitorResults(serpKey);
      console.log(`   ✅ ${results.serp_results.competitors.length} resultat`);
    } catch (e) {
      console.error(`   ❌ ${e.message}`);
      results.serp_results.competitors = [];
    }
  }

  // Perplexity: deep analysis
  const queries = buildQueries(brand);
  console.log(`\n🤖 Researchar via Perplexity (${queries.length} frågor)...`);

  // Build context from SERP results to include in analysis
  const serpContext = serpKey && results.serp_results.competitors?.length
    ? `\n\nGooglesökresultat att utgå ifrån:\n${results.serp_results.competitors.slice(0, 8).map(r => `- ${r.title}: ${r.snippet} (${r.link})`).join("\n")}`
    : "";

  for (let i = 0; i < queries.length; i++) {
    const q = queries[i];
    console.log(`\n[${i + 1}/${queries.length}] ${q.label}...`);
    try {
      const answer = await askPerplexity(q.query + serpContext, perplexityKey);
      results.research[q.id] = { label: q.label, query: q.query, answer };
      console.log(`   ✅ ${answer.length} tecken`);
    } catch (e) {
      console.error(`   ❌ ${e.message}`);
      results.research[q.id] = { label: q.label, query: q.query, answer: `FEL: ${e.message}` };
    }
    // Small delay to avoid rate limiting
    if (i < queries.length - 1) await new Promise(r => setTimeout(r, 800));
  }

  mkdirSync(join(ROOT, ".tmp"), { recursive: true });
  writeFileSync(OUTPUT_FILE, JSON.stringify(results, null, 2), "utf-8");
  console.log(`\n✅ Research sparad: ${OUTPUT_FILE}`);
  console.log("   Kör nu: node tools/generate_report.js");
}

main().catch(e => { console.error("Fel:", e); process.exit(1); });
